lsp/nls/src/field_walker.rs

```diff
@@ -13,35 +13,55 @@ use nickel_lang_core::{
 
 use crate::{identifier::LocIdent, requests::completion::CompletionItem, server::Server};
 
-/// A `FieldHaver` is something that... has fields.
+/// A `Container` is something that has elements.
 ///
-/// You can use a [`FieldResolver`] to resolve terms, or terms with paths, to `FieldHaver`s.
+/// The elements could have names (e.g. in a record) or not (e.g. in an array).
+///
+/// You can use a [`FieldResolver`] to resolve terms, or terms with paths, to `Container`s.
 #[derive(Clone, Debug, PartialEq)]
-pub enum FieldHaver {
+pub enum Container {
```

Comment:
I wonder: would that make sense to make this a trait instead? It seems that most of what you do below is to match on the type of container, and then apply specific behavior. It sounds a lot like what traits are for.

Also, it feels slightly odd to mix array and records, leading to a rather vague notion of container (I guess it's why this is hard to give it a meaningful name) and an `EltId` which also feels like it smashes two different things together to then match on the different possibilities and apply specific behavior (also, if I'm understanding correctly, some illegal cases such as trying to get an ident from an array aren't statically excluded but just return `None`).

At this point I might be oblivious to important details that make the current approach actually justified, sorry if this is the case. But from a distance, would that lead to much code duplication to have:

A trait `FieldHaver` with the methods you already implemented before this PR, and implement it for `RecordData`, `Type`, and `RecordRows`.
A trait `Indexable` (that's a bad name and it doesn't relate to `FieldHaver`, so it's doubly bad...) that is mostly doing the same but for array elements, which doesn't use `Ident` nor `field` in the method's name, and implement that for `Array`.

What do you think?

Reply:
Good points, I will think more about this over the break. I think there is at least one situation where you do want to dynamically handle either a record field or an array element, but it's probably a smaller use case and there's probably a better organization.

To elaborate, suppose we're trying to complete at the `a` in `{foo = [{ a }]} | {foo | Array { aardvark | String }}`. The way we currently do this without any evaluation is to first construct the path to the place we want completion. In some weird notation I just made up, this path looks like `. foo / !!` (where the `!!` means an array index -- we don't track what the index actually was). Then we follow the same path on the contract `{foo | Array { aardvark | String }}` to find the `aardvark` completion. 

So the path does need to alternate at runtime between field accesses and array indices, and it might also happen that the path-follower tries to match an array index against a record (say, if the contract had been `{foo | {_: {aardvark | String}}}`). I don't know of any *valid* nickel code where this will happen, but of course nls also needs to handle invalid nickel code.